lib/Driver/ToolChains.cpp

```diff
@@ -526,6 +353,186 @@ ToolChain::constructInvocation(const CompileJobAction &job,
   return II;
 }
 
+const char *ToolChain::JobContext::computeFrontendModeForCompile() const {
+  switch (OI.CompilerMode) {
+  case OutputInfo::Mode::StandardCompile:
+  case OutputInfo::Mode::SingleCompile:
+  case OutputInfo::Mode::BatchModeCompile:
+    break;
+  case OutputInfo::Mode::Immediate:
+  case OutputInfo::Mode::REPL:
+    llvm_unreachable("REPL and immediate modes handled elsewhere");
+  }
+  switch (Output.getPrimaryOutputType()) {
+  case types::TY_Object:
+    return "-c";
+    break;
+  case types::TY_PCH:
+    return "-emit-pch";
+    break;
+  case types::TY_RawSIL:
+    return "-emit-silgen";
+    break;
+  case types::TY_SIL:
+    return "-emit-sil";
+    break;
+  case types::TY_RawSIB:
+    return "-emit-sibgen";
+    break;
+  case types::TY_SIB:
+    return "-emit-sib";
+    break;
+  case types::TY_LLVM_IR:
+    return "-emit-ir";
+    break;
+  case types::TY_LLVM_BC:
+    return "-emit-bc";
+    break;
+  case types::TY_Assembly:
+    return "-S";
+    break;
+  case types::TY_SwiftModuleFile:
+    // Since this is our primary output, we need to specify the option here.
+    return "-emit-module";
+    break;
+  case types::TY_ImportedModules:
+    return "-emit-imported-modules";
+    break;
+  case types::TY_IndexData:
+    return "-typecheck";
+    break;
+  case types::TY_Remapping:
+    return "-update-code";
+    break;
+  case types::TY_Nothing:
+    // We were told to output nothing, so get the last mode option and use that.
+    if (const Arg *A = Args.getLastArg(options::OPT_modes_Group))
+      return A->getSpelling().data();
+    else
+      llvm_unreachable("We were told to perform a standard compile, "
+                       "but no mode option was passed to the driver.");
+    break;
+  case types::TY_Swift:
+  case types::TY_dSYM:
+  case types::TY_AutolinkFile:
+  case types::TY_Dependencies:
+  case types::TY_SwiftModuleDocFile:
+  case types::TY_ClangModuleFile:
+  case types::TY_SerializedDiagnostics:
+  case types::TY_ObjCHeader:
+  case types::TY_Image:
+  case types::TY_SwiftDeps:
+  case types::TY_ModuleTrace:
+  case types::TY_TBD:
+  case types::TY_OptRecord:
+    llvm_unreachable("Output type can never be primary output.");
+  case types::TY_INVALID:
+    llvm_unreachable("Invalid type ID");
+  }
+}
+
+void ToolChain::JobContext::addCompileArgumentsEitherOnCommandLineOrInFilelists(
```

Comment:
This definitely does not belong on JobContext. It is not JobContext's job (heh) to populate argument lists, and (same as above) I don't think it makes sense for JobContext to know about frontend options at all.

Reply:
Let's talk about this offline. We can hash out the utility of defining JobContext's job according to various criteria. There's something I don't yet understand going on here.

Reply:
Move to CompileJobAction, or where??